tools/halide_image.h

```diff
@@ -1,278 +1,696 @@
-// This header defines a simple Image class which wraps a buffer_t. This is
-// useful when interacting with a statically-compiled Halide pipeline emitted by
-// Func::compile_to_file, when you do not want to link your processing program
-// against Halide.h/libHalide.a.
+// A simple Image class which wraps a buffer_t. This is useful
+// when interacting with a statically-compiled Halide pipeline emitted
+// by Func::compile_to_file, when you do not want to link your
+// processing program against Halide.h/libHalide.a.
 
-#ifndef HALIDE_TOOLS_IMAGE_H
-#define HALIDE_TOOLS_IMAGE_H
+#ifndef HALIDE_RUNTIME_IMAGE_H
+#define HALIDE_RUNTIME_IMAGE_H
 
-#include <cassert>
-#include <cstdlib>
-#include <limits>
 #include <memory>
 #include <vector>
-#include <stdint.h>  // <cstdint> requires C++11
+#include <cassert>
+#include <stdint.h>
+#include <string.h>
 
 #include "HalideRuntime.h"
 
 namespace Halide {
 namespace Tools {
 
-template<typename T>
-class Image {
-    struct Contents {
-        Contents(const buffer_t &b, uint8_t *a) : buf(b), ref_count(1), alloc(a) {}
-        buffer_t buf;
-        int ref_count;
-        uint8_t *alloc;
+template<typename Fn>
+void for_each_element(const buffer_t &buf, Fn f);
+
+/** A templated Image class derived from buffer_t that adds
+ * functionality. T is the element type, and D is the maximum number
+ * of dimensions. It must be less than or equal to 4 until we have
+ * support for higher-dimensional buffers. */
+template<typename T, int D = 4, void *(*Allocate)(size_t) = malloc, void (*Deallocate)(void *) = free>
```

Comment:
Why is this using malloc/free instead of halide_malloc/halide_free? (E.g. halide_malloc provides stronger alignment guarantees.) Obviously an adapter will be required to handle user_context, but perhaps that is a sign we have an issue with the design.